### ISSUE TITLE
CI: add GitHub Actions check for Apple Silicon (M1) build

### ISSUE THREAD

**Issue by 678098** [bloomberg/blazingmq#116]
### Is there an existing proposal for this?

- [X] I have searched the existing proposals

### Is your feature request related to a problem?

We don't have automatic checks for M1/M2 builds.

### Describe the solution you'd like

According to this roadmap, Apple Silicon (M1) is in public beta in GitHub Actions:

https://github.com/github/roadmap/issues/528

Might be possible to use the corresponding runner to check darwin build script automatically.

**Comment by Sudo-Guy** [comment]
I can pick this up. Just to clarify we want to take advantage of the M1 runners to see if blazingmq builds via the darwin build script successfully?

**Comment by 678098** [comment]
> I can pick this up. Just to clarify we want to take advantage of the M1 runners to see if blazingmq builds via the darwin build script successfully?

Hi @Sudo-Guy! That is correct, need to implement a check for the darwin build script using M1 or M2 (if available) GitHub Runner.
Will assign it to you then. Thanks!

**Comment by Sudo-Guy** [comment]
I'll get around to it this weekend. Minor issue is that I don't have access to a ARM mac machine right now, nor are free runners currently available.

**Comment by kgantchev** [comment]
> ...if free Darwin runners are available.

Might I make a suggestion? Feel free to use [FlyCI](https://www.flyci.net/)'s M1 and M2 runners. Our runners are on average 2x faster and 2x cheaper than GitHub's **AND** we have a free tier for OSS projects (see below).

## Install Instructrions 

Easily replace your M1 runners: 
```diff
jobs:
 ci:
-    runs-on: macos-latest
+    runs-on: flyci-macos-large-latest-m1
   steps:
   - name: 👀 Checkout repo
     uses: actions/checkout@v4
```

Or try the M2 runners: 
```diff
jobs:
  ci:
-    runs-on: macos-latest
+    runs-on: flyci-macos-large-latest-m2
    steps:
      - name: 👀 Checkout repo
        uses: actions/checkout@v4
 ```

## Pricing  

| Processor | vCPU | RAM (GB) | Storage | Label | Price on FlyCI | Price on GitHub |
|-----------|-------|-----------|----------|-------------------------------|----------------|-----------------|
| M1        | 4    | 7        | 28 GB | flyci-macos-large-latest-m1   | $0.06          | -             |
| M1        | 8    | 14       | 28 GB | flyci-macos-xlarge-latest-m1  | $0.12          | $0.16           |
| M2        | 4    | 7        | 28 GB | flyci-macos-large-latest-m2   | $0.08          | -             |
| M2        | 8    | 14       | 28 GB | flyci-macos-xlarge-latest-m2  | $0.16          | -             |

## 500 mins/month Free for Public Repos

If your repo is public, then FlyCI offers 500 mins/month of free M1 runner usage with the `flyci-macos-large-latest-m1` runner.  

Best Regards,
Kiril Gantchev
CEO and co-founder of [FlyCI](https://flyci.net)

**Comment by 678098** [comment]
Hi @kgantchev! Thank you for the suggestion. The proposed option is great for small open source projects. However, we have many developers who actively make changes in the repo.

Some calculations. The free quota is 500 min, consider M1/M2 runner time 20 min for each CI launch. With these assumptions, the quota will be exceeded after ~25 CI launches.

This means:
- On PR checks - exceed the quota, because we have >25 PR updates per month.
- Daily main branch checks - exceed the quota, because 25 is less than a number of days in month.
- Weekly main branch checks - fit the quota.

In conclusion, weekly checks could be performed, which could be good enough for us. I will discuss it with the team.

**Comment by 678098** [comment]
> GitHub Actions: Introducing the new M1 macOS runner available to open source!

https://github.blog/changelog/2024-01-30-github-actions-introducing-the-new-m1-macos-runner-available-to-open-source/

**Comment by 678098** [comment]
Hi @syuzvinsky, could you do this when you have time?

**Comment by pniedzielski** [comment]
For reference, [this is how memray has done it](https://github.com/bloomberg/memray/commit/dcf747db84a07fdd21c0489c59c95b59f19aea0c).

**Comment by 678098** [comment]
https://github.com/bloomberg/blazingmq/pull/205
